### DISCUSSION TITLE
Stateful incremental score calculation

### DISCUSSION THREAD

**Discussion by lightoze** [TimefoldAI/timefold-solver#1225] [General]
I have a planning problem with a seemingly simple model, but with a twist, and will be very thankful if someone gives an advice his to apply Timefold to it. 

I have an initial system state and a number of actions (planning items, a few hundreds of them) which can be applied to it, in order. Each item can be picked only once, but not all of them need to be selected and domain knowledge says that normally less than 50% will be in the correct solution. So it seems like planning list variable should work for me, but constraint/score calculation complicates everything. 

Each time an item is applied, it changes the system state, and soft/hard score it adds to additive cumulative result depends on this state. The state is rather complicated and I want to find a way to do this incrementally. Otherwise recalculating the state and score for a given set of items will become way more time consuming. How is it possible to achieve that? 

If constraint API was guaranteed to be ordered for members of the same planning list variable, and there was a folding operation, my task would be doable with that, but there's no such possibility, right?

**Comment by Christopher-Chianelli** [comment]
- A `@PlanningListVariable` is ordered. Since values can be unassigned, you want `@PlanningListVariable(allowUnassignedValues=true)`.
- You can use a [@CascadingUpdateShadowVariable](https://docs.timefold.ai/timefold-solver/latest/using-timefold-solver/modeling-planning-problems#tailChainVariable) to keep track of the state.

**Comment by lightoze** [comment]
So I'll keep the whole system's state in the item's shadow variable and update it based on the previous one. I'll try that, thanks!
